From 4bd7de9be20e570aa01727b0a8b5b8c14597f1f5 Mon Sep 17 00:00:00 2001
From: Markus Boehme <markus.boehme@chainguard.dev>
Date: Mon, 5 Jan 2026 19:41:54 +0100
Subject: [PATCH] container-selinux: add stubs for all container interfaces

We dropped the container-selinux policy for undesired effects in our
selinux-policy package [1]. However, the Fedora policy assumes it to be
present and contains references to interfaces provided by the
container-selinux policy. Trying to compile the policy without those
results in syntax errors, e.g.:

    policy/modules/kernel/domain.te:749:ERROR 'syntax error' at token \
    	'container_spc_stream_connect' on line 176683: \
	container_spc_stream_connect(domain)

Fix this by adding stubs for all interfaces defined by the
container-selinux policy/project. The stub file was generated from the
original container.if interface definitions by preserving comments,
newlines, and the start of the interface definition. Since the body of
the resulting SELinux policy "optional" block may not be empty, add a
dependency on a non-existing type. This results in the block being
removed entirely during policy compilation via the "optional_policy"
macro.

I generated the interface stubs using:

    grep -E -e '^\s*$' -e '^#' -e '^interface' container.if \
    	| uniq \
	| perl -ne 'print; printf "\tgen_require(`type does_not_exist_t;'\'')\n'\'')\n" if /^interface/' \
	>container-stub.if

[1]: https://github.com/chainguard-dev/enterprise-packages/pull/43588

Signed-off-by: Markus Boehme <markus.boehme@chainguard.dev>
---
 policy/modules/contrib/container-stub.if | 671 +++++++++++++++++++++++
 1 file changed, 671 insertions(+)
 create mode 100644 policy/modules/contrib/container-stub.if

diff --git a/policy/modules/contrib/container-stub.if b/policy/modules/contrib/container-stub.if
new file mode 100644
index 000000000..50b406563
--- /dev/null
+++ b/policy/modules/contrib/container-stub.if
@@ -0,0 +1,671 @@
+## <summary>The open-source application container engine.</summary>
+
+########################################
+## <summary>
+##	Execute container in the container domain.
+## </summary>
+## <param name="domain">
+## <summary>
+##	Domain allowed to transition.
+## </summary>
+## </param>
+#
+interface(`container_runtime_domtrans',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container runtime in the container runtime domain
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+## <param name="role">
+##	<summary>
+##	Role allowed access.
+##	</summary>
+## </param>
+## <rolecap/>
+#
+interface(`container_runtime_run',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container in the caller domain.
+## </summary>
+## <param name="domain">
+## <summary>
+##	Domain allowed to transition.
+## </summary>
+## </param>
+#
+interface(`container_runtime_exec',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read the process state of container runtime
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_read_state',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Write to /proc/PID of container runtime.
+##	This is needed e.g. to set uid_map or gid_map
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_write_proc_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Search container lib directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_search_lib',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container lib directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_exec_lib',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read container lib files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_read_lib_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read container share files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_read_share_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read container runtime tmpfs files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_runtime_read_tmpfs_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container share files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_share_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container share dirs.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_share_dirs',`
+	gen_require(`type does_not_exist_t;')
+')
+
+######################################
+## <summary>
+##	Allow the specified domain to execute container shared files
+##	in the caller domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_exec_share_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container config files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_config_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container lib files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_lib_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_dirs',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Manage container lib directories.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_manage_lib_dirs',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Create objects in a container var lib directory
+##	with an automatic type transition to
+##	a specified private type.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+## <param name="private_type">
+##	<summary>
+##	The type of the object to create.
+##	</summary>
+## </param>
+## <param name="object_class">
+##	<summary>
+##	The class of the object to be created.
+##	</summary>
+## </param>
+## <param name="name" optional="true">
+##	<summary>
+##	The name of the object being created.
+##	</summary>
+## </param>
+#
+interface(`container_lib_filetrans',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read container PID files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_read_pid_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container server in the container domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`container_systemctl',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read and write container shared memory.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_rw_sem',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Allow the specified domain to append
+##	to container files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_append_file',`
+	gen_require(`type does_not_exist_t;')
+')
+
+#######################################
+## <summary>
+##  Read and write the container pty type.
+## </summary>
+## <param name="domain">
+##  <summary>
+##  Domain allowed access.
+##  </summary>
+## </param>
+#
+interface(`container_use_ptys',`
+	gen_require(`type does_not_exist_t;')
+')
+
+#######################################
+## <summary>
+##      Allow domain to create container content
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`container_filetrans_named_content',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Connect to container over a unix stream socket.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Connect to SPC containers over a unix stream socket.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_spc_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	All of the rules required to administrate
+##	an container environment
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_admin',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container_auth_exec_t in the container_auth domain.
+## </summary>
+## <param name="domain">
+## <summary>
+##	Domain allowed to transition.
+## </summary>
+## </param>
+#
+interface(`container_auth_domtrans',`
+	gen_require(`type does_not_exist_t;')
+')
+
+######################################
+## <summary>
+##	Execute container_auth in the caller domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_auth_exec',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Connect to container_auth over a unix stream socket.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_auth_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	container domain typebounds calling domain.
+## </summary>
+## <param name="domain">
+## <summary>
+##	Domain to be typebound.
+## </summary>
+## </param>
+#
+interface(`container_runtime_typebounds',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Allow any container_runtime_exec_t to be an entrypoint of this domain
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+## <rolecap/>
+#
+interface(`container_runtime_entrypoint',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_exec_lib',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_read_share_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_exec_share_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_manage_lib_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_manage_lib_dirs',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_lib_filetrans',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_read_pid_files',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_systemctl',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_use_ptys',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+interface(`docker_spc_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Read the process state of spc containers
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_spc_read_state',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Creates types and rules for a basic
+##	container runtime process domain.
+## </summary>
+## <param name="prefix">
+##	<summary>
+##	Prefix for the domain.
+##	</summary>
+## </param>
+#
+
+########################################
+## <summary>
+##	Creates types and rules for a basic
+##	container process domain.
+## </summary>
+## <param name="prefix">
+##	<summary>
+##	Prefix for the domain.
+##	</summary>
+## </param>
+## <param name="prefix">
+##	<summary>
+##	Prefix for the file type.
+##	</summary>
+## </param>
+#
+
+########################################
+## <summary>
+##	Manage container files template
+## </summary>
+## <param name="prefix">
+##	<summary>
+##	Prefix for the domain.
+##	</summary>
+## </param>
+## <param name="prefix">
+##	<summary>
+##	Prefix for the file type.
+##	</summary>
+## </param>
+#
+
+########################################
+## <summary>
+##	Read and write a spc_t unnamed pipe.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_spc_rw_pipes',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute container in the container domain.
+## </summary>
+## <param name="domain">
+## <summary>
+##	Domain allowed to transition.
+## </summary>
+## </param>
+#
+interface(`container_kubelet_domtrans',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Execute kubelet_exec_t in the kubelet_t domain
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+## <param name="role">
+##	<summary>
+##	Role allowed access.
+##	</summary>
+## </param>
+## <rolecap/>
+#
+interface(`container_kubelet_run',`
+	gen_require(`type does_not_exist_t;')
+')
+
+########################################
+## <summary>
+##	Connect to kubelet over a unix stream socket.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`container_kubelet_stream_connect',`
+	gen_require(`type does_not_exist_t;')
+')
+
+#######################################
+## <summary>
+##      Create a file type used for container files.
+## </summary>
+## <param name="script_file">
+##      <summary>
+##      Type to be used for an container file.
+##      </summary>
+## </param>
+#
+interface(`container_file',`
+	gen_require(`type does_not_exist_t;')
+')
+
-- 
2.52.0

